modules/core/src/main/scala/org/tessellation/domain/snapshot/programs/Download.scala

```diff
@@ -57,20 +58,32 @@ object Download {
     val minBatchSizeToStartObserving: Long = 1L
     val observationOffset = NonNegLong(4L)
     val fetchSnapshotDelayBetweenTrials = 10.seconds
+    val checkJoinablePeersDelayBetweenAttempts = 1.minute
+    val retryCheckJoinablePeersMax = 10.minutes
 
     type DownloadResult = (Signed[GlobalIncrementalSnapshot], GlobalSnapshotInfo)
     type ObservationLimit = SnapshotOrdinal
 
-    def download: F[Unit] =
-      nodeStorage
-        .tryModifyState(NodeState.WaitingForDownload, NodeState.DownloadInProgress, NodeState.WaitingForObserving)(start)
-        .flatMap(observe)
-        .flatMap { result =>
-          val ((snapshot, context), observationLimit) = result
+    def waitForPeers(minPeers: Int): F[Set[Peer]] = retryingOnFailures[Set[Peer]](
+      limitRetriesByCumulativeDelay(
+        threshold = retryCheckJoinablePeersMax,
+        policy = constantDelay(checkJoinablePeersDelayBetweenAttempts)
+      ),
+      a => (a.size >= minPeers).pure[F],
+      (a, details) => logger.info(s"Discovered ${a.size} selectable peers, waiting for more: $details")
+    )(clusterStorage.getResponsivePeers.map(_.filter(_.state === Ready)))
 
-          consensus.manager.startFacilitatingAfterDownload(observationLimit, snapshot, context)
-        }
-        .onError(logger.error(_)("Unexpected failure during download!"))
+    def download: F[Unit] =
+      waitForPeers(2) >> waitForPeers(1) >>
```

Comment:
if `waitForPeers(2)` would succeed, why we have to again `waitForPeers(1)`?

Reply:
The function `waitForPeers` has a `10 minutes` limit. Afterwards, it continues regardless of whether the check passed or not. So, it will move on from `waitForPeers(2)` when either:

- There are two or more discovered, selectable peers.
- 10 minutes have elapsed.

Since, `PeerSelect` needs, at minimum, one discovered, selectable peer, we have to run `waitForPeers(1)` to wait to discover one selectable peer. If `waitForPeers(1)` times out, we'll move to `PeerSelect`, which will throw `NoPeersToSelect`.

Reply:
Ok, I understand how it should work but it's weird that we need to run both even if `waitForPeers(2)` succeeds.
IMO it should be a fallback to one peer instead of sequentially chaining these operations

Reply:
If you have any guidance on a better way to do this, I'm all ears. This is the best I can do right now.

Reply:
```scala
def waitForPeers(n: Long): EitherT[F, NoPeers, Unit] = ???

waitForPeers(99).orElse(waitForPeers(2)).flatMap {
  case Right(_) => download
  case Left(_)  => throw no peers and stop
}
```
(this is just the pseudocode, don't copy it 1:1)

Reply:
Changed to

```
    def waitForPeers: F[Unit] = {
      val peers = if (appEnvironment === Dev) {
        waitUntil(fallbackPeersCount)
      } else {
        waitUntil(minPeersCount) >>
          Async[F].sleep(additionalDiscoveryDelay) >>
          waitUntil(fallbackPeersCount)
      }

      peers.flatMap {
        case s if s.nonEmpty => ().pure
        case _               => NoPeersToSelect.raiseError[F, Unit]
      }
    }
```

Reply:
@cngo-github what has been changed? I see no logical difference between this and the previous code, still `waitUntil(fallbackPeersCount)` is executed even if `minPeersCount` succeeded. Also, `sleep` is executed always, even if `minPeersCount` has been reached immediately.

Reply:
@marcinwadon , I believe that according to our offline discussion, this is no longer an issue.

Reply:
IMO the case is still valid. The additional delay should be there as we discussed, but `fallbackPeersCount` will be executed one way or another, but IMO it shouldn't if the first `waitUntil` succeeds and there are more peers already found

Reply:
So, the purpose of

```
waitUntil(minPeersCount) >>
          Async[F].sleep(additionalDiscoveryDelay)
```

is simply to wait for more than 1 peer. Since we don't care about the exact peer count, that has been simplified to just a single delay. Afterwards, we check for 1 peer since `PeerSelect` needs at least 1 peer.

TLDR:

The function has been simplified to

```
    def waitForPeers: F[Unit] =
      Async[F].sleep(additionalDiscoveryDelay).unlessA(appEnvironment === Dev) >>
        waitUntil(fallbackPeersCount).flatMap {
          case s if s.nonEmpty => ().pure
          case _               => NoPeersToSelect.raiseError[F, Unit]
        }
```